src/macos/cpid_macos.c

```diff
@@ -0,0 +1,288 @@
+    EVP_MD_CTX *digest_context;
+    EVP_MD *sha256;
+    uint8_t digest_destination_buffer[EVP_MAX_MD_SIZE];
+    digest_input_content_t digest_input_content;
+} *cpid_handle_t;
+
+_Static_assert(EVP_MAX_MD_SIZE >= SHA256_BUFFER_SIZE, "OpenSSL EVP_MAX_MD_SIZE must be larger that SHA256_BUFFER_SIZE.");
+_Static_assert(SHA256_BUFFER_SIZE >= sizeof(uuid_t), "SHA256_BUFFER_SIZE must be larger that uuid_t.");
+
+static int cpid_get_serial_number(char * serial_number, size_t serial_number_size) {
+    if(NULL == serial_number || 0 == serial_number_size) {
+        return RETURN_ERROR;
+    }
+
+    io_service_t platformExpert = IOServiceGetMatchingService(kIOMainPortDefault, IOServiceMatching("IOPlatformExpertDevice"));
+    if (MACH_PORT_NULL == platformExpert) {
+        return RETURN_ERROR;
+    }
+
+    CFStringRef serialNumberAsCFString = (CFStringRef) IORegistryEntryCreateCFProperty(platformExpert, CFSTR(kIOPlatformSerialNumberKey), kCFAllocatorDefault, 0);
+    IOObjectRelease(platformExpert);
+    if (NULL == serialNumberAsCFString) {
+        return RETURN_ERROR;
+    }
+
+    Boolean get_cstring_succeeded = CFStringGetCString(serialNumberAsCFString, serial_number, serial_number_size, kCFStringEncodingUTF8);
+    CFRelease(serialNumberAsCFString);
+    if(!get_cstring_succeeded) {
+        return RETURN_ERROR;
+    }
+
+    return RETURN_SUCCESS;
+}
+
+static int cpid_get_hardware_uuid(uuid_t hardware_uuid) {
+    if(NULL == hardware_uuid) {
+        return RETURN_ERROR;
+    }
+
+    uuid_string_t uuid_receive_text_buffer = {0};
+
+    io_registry_entry_t ioRegistryRoot = IORegistryEntryFromPath(kIOMainPortDefault, "IOService:/");
+    if(MACH_PORT_NULL == ioRegistryRoot) {
+        return RETURN_ERROR;
+    }
+
+    CFStringRef uuidCf = (CFStringRef) IORegistryEntryCreateCFProperty(ioRegistryRoot, CFSTR(kIOPlatformUUIDKey), kCFAllocatorDefault, 0);
+    IOObjectRelease(ioRegistryRoot);
+    if (NULL == uuidCf) {
+        return RETURN_ERROR;
+    }
+    
+    Boolean get_cstring_succeeded = CFStringGetCString(uuidCf, uuid_receive_text_buffer, sizeof(uuid_string_t), kCFStringEncodingMacRoman);
+    CFRelease(uuidCf);
+    if(!get_cstring_succeeded) {
+        return RETURN_ERROR;
+    }
+    
+    int return_code = uuid_parse(uuid_receive_text_buffer, hardware_uuid);
+    if(0 != return_code) {
+        return RETURN_ERROR;
+    }
+
+    return RETURN_SUCCESS;
+}
+
+static int cpid_get_process_creation_time(uint32_t pid, process_creation_time_t *process_creation_time) {
+    // pid check needed for conversion to signed below
+    if(pid > MAX_PID || NULL == process_creation_time) {
+        return RETURN_ERROR;
+    }
+
+    int mib[4] = {CTL_KERN, KERN_PROC, KERN_PROC_PID, (int) pid};
+
+    struct kinfo_proc process_info = {0};
+    size_t info_size = sizeof(process_info);
+
+    int return_code = sysctl(mib, 4, &process_info, &info_size, NULL, 0);
+    if (0 != return_code) {
+        return RETURN_ERROR;
+    }
```

Comment:
return_code doesn't seem to be used elsewhere and isn't used as a "return_code" in the context of this funciton. I'd suggest removing it.

```suggestion
    if (0 != sysctl(mib, 4, &process_info, &info_size, NULL, 0)) {
        return RETURN_ERROR;
    }
```

Reply:
There is similar code in other functions in this file

Reply:
Yeah I get it will be optimized away. If you're at the point of step through debugging it is easy enough to break out the information you want if needed. To me defining a variable for it it to be basically unused seems verbose.

Certainly not a deal breaker either way.